### PR TITLE
feat: Update model architecture and multiprocessing

### PR DESCRIPTION
**Pull request by mihirkatare**
- Add multiprocessing support for the dataloader (cache-based multiprocessing)
- Add residual network architectures 
- Separate model and networks for scalability
- Add draft of API arguments to README

### CODE REVIEW

File: model.py [hunk 1]
```diff
@@ -11,33 +11,9 @@
...
-        y = self.relu(self.fc1(x))
-        y = self.relu(self.fc2(y))
-        y = self.relu(self.fc3(y))
-        y = self.relu(self.fc4(y))
-        y = self.relu(self.fc5(y))
-        y = self.selu(self.fc6(y))
-        return y
+import hist
+from sklearn.metrics import mean_absolute_percentage_error as mape
+from networks import *
```
Comment (matthewfeickert) [inline model.py:16]: In general we should always avoid doing `import *` as this basically totally destroys our namespace and makes it much more difficult to understand how changes in APIs propagate in the future. Can we explicitly give the imports here?

File: model.py [hunk 2]
```diff
@@ -106,15 +82,21 @@ def inference(self, device):
...
-        plt.hist(test_y_pred.detach().cpu().numpy(), bins=nbins, histtype = "step", label = "DNN Prediction")
-        plt.xlabel("-log_10(DY weight)")
-        plt.ylabel("events")
-        plt.legend()
+        # plt.hist(utils.test_Y.numpy(), bins=nbins, histtype = "step", color = "r", label = "Test Dataset")
+        # plt.hist(test_y_pred.detach().cpu().numpy(), bins=nbins, histtype = "step", label = "DNN Prediction")
+        # plt.xlabel("-log_10(DY weight)")
+        # plt.ylabel("events")
+        # plt.legend()
+        # plt.savefig(self.manager.args["save_testing_histogram_at"])
```
Comment (matthewfeickert) [inline model.py:92]:
```suggestion
```

Probably better to remove code rather than have commended out code in version control.

File: model.py [hunk 3]
```diff
@@ -106,15 +82,21 @@ def inference(self, device):
...
+        hist_1 = hist.Hist(hist.axis.Regular(nbins, 0, 10, name="-log_10(DY weight)")).fill(utils.test_Y.numpy())
+        hist_2 = hist.Hist(hist.axis.Regular(nbins, 0, 10, name="-log_10(DY weight)")).fill(test_y_pred.detach().cpu().numpy())
+        fig = plt.figure(figsize=(10, 8))
+        main_ax_artists, sublot_ax_arists = hist_1.plot_ratio(hist_2, rp_ylabel=r"Ratio", rp_num_label="Test Dataset", rp_denom_label="DNN Prediction", rp_uncert_draw_type="line")
+        plt.yticks(np.arange(-5,6))
+        plt.grid()
```
Comment (matthewfeickert) [inline model.py:98]: This is fine for now, but we can go and specify which axis objects we want this applied to later on rather than using the `plt` API.

**Review by matthewfeickert — CHANGES_REQUESTED** [review]
@mihir-katare This looks good overall. :+1: I've put some small comments in.

Do we need to keep the contents of `post/` in version control? They're all binaries, so unless there is a necessity to have them be versioned I'd suggest instead attaching them as artifacts to a GitHub release.

**Comment by mihirkatare** [comment]
> @mihirkatare Do we need to keep the contents of `post/` in version control? They're all binaries, so unless there is a necessity to have them be versioned I'd suggest instead attaching them as artifacts to a GitHub release (which for the time being would mean copying out elsewhere to keep).

We mention that the library will be pip-installable. Does that include the pre-trained models? i.e the .pth binaries

**Comment by mihirkatare** [comment]
Okay the binaries have been removed from main
